File: tests/corpus/concept.java
// vyakarana stand-in -- not from vidya. See docs/adr/0006-standin-corpus-policy.md.
// Re-sync when vidya adds a Java reference sample.
//
// Lexing + parsing of arithmetic expressions. Mirrors the
// content/lexing_and_parsing/ pattern (rust.rs, go.go, etc.).

package com.example.vyakarana;

import java.util.ArrayList;
import java.util.List;

/*
 * Multi-line block comment exercising the slash-star pair rule
 * across multiple lines. C/C++/Java family share this shape.
 */

public final class ConceptParser {

    // ── Tokens ───────────────────────────────────────────────────────

    public enum TokenKind {
        NUMBER, PLUS, MINUS, STAR, SLASH, LPAREN, RPAREN, EOF
    }

    public static final class Token {
        public final TokenKind kind;
        public final String text;
        public final int pos;

        public Token(TokenKind kind, String text, int pos) {
            this.kind = kind;
            this.text = text;
            this.pos = pos;
        }

        @Override
        public String toString() {
            return String.format("Token(%s, %s, %d)", kind, text, pos);
        }
    }

    // ── Lexer ────────────────────────────────────────────────────────

    public static final class Lexer {
        private final String source;
        private int pos = 0;

        public Lexer(String source) {
            this.source = source;
        }

        private char peek() {
            return pos < source.length() ? source.charAt(pos) : '\0';
        }

        private void skipWhitespace() {
            while (pos < source.length() && Character.isWhitespace(source.charAt(pos))) {
                pos++;
            }
        }

        public Token next() {
            skipWhitespace();
            int start = pos;

            if (pos >= source.length()) {
                return new Token(TokenKind.EOF, "", pos);
            }

            char c = source.charAt(pos);

            switch (c) {
                case '+': pos++; return new Token(TokenKind.PLUS,   "+", start);
                case '-': pos++; return new Token(TokenKind.MINUS,  "-", start);
                case '*': pos++; return new Token(TokenKind.STAR,   "*", start);
                case '/': pos++; return new Token(TokenKind.SLASH,  "/", start);
                case '(': pos++; return new Token(TokenKind.LPAREN, "(", start);
                case ')': pos++; return new Token(TokenKind.RPAREN, ")", start);
                default:
                    break;
            }

            if (Character.isDigit(c)) {
                while (pos < source.length() && Character.isDigit(source.charAt(pos))) {
                    pos++;
                }
                String text = source.substring(start, pos);
                return new Token(TokenKind.NUMBER, text, start);
            }

            throw new IllegalStateException(
                String.format("unexpected character '%c' at %d", c, pos));
        }
    }

    // ── Pratt parser ─────────────────────────────────────────────────

    public sealed interface Expr permits NumberExpr, BinOpExpr, UnaryExpr {}
    public record NumberExpr(long value) implements Expr {}
    public record BinOpExpr(char op, Expr left, Expr right) implements Expr {}
    public record UnaryExpr(char op, Expr operand) implements Expr {}

    public static final class Parser {
        private final Lexer lexer;
        private Token current;

        public Parser(String source) {
            this.lexer = new Lexer(source);
            this.current = lexer.next();
        }

        private Token advance() {
            Token prev = current;
            current = lexer.next();
            return prev;
        }

        private static int infixBp(TokenKind k) {
            return switch (k) {
                case PLUS, MINUS -> 1;
                case STAR, SLASH -> 3;
                default -> 0;
            };
        }

        public Expr parseExpr(int minBp) {
            Expr left = parsePrimary();
            while (true) {
                int bp = infixBp(current.kind);
                if (bp == 0 || bp < minBp) break;
                char op = current.text.charAt(0);
                advance();
                Expr right = parseExpr(bp + 1);
                left = new BinOpExpr(op, left, right);
            }
            return left;
        }

        private Expr parsePrimary() {
            if (current.kind == TokenKind.MINUS) {
                advance();
                return new UnaryExpr('-', parsePrimary());
            }
            if (current.kind == TokenKind.LPAREN) {
                advance();
                Expr inner = parseExpr(0);
                if (current.kind != TokenKind.RPAREN) {
                    throw new IllegalStateException("expected )");
                }
                advance();
                return inner;
            }
            if (current.kind == TokenKind.NUMBER) {
                long val = Long.parseLong(current.text);
                advance();
                return new NumberExpr(val);
            }
            throw new IllegalStateException("unexpected token " + current);
        }
    }

    // ── Demo ─────────────────────────────────────────────────────────

    public static long eval(Expr e) {
        return switch (e) {
            case NumberExpr n  -> n.value();
            case UnaryExpr u   -> -eval(u.operand());
            case BinOpExpr b   -> switch (b.op()) {
                case '+' -> eval(b.left()) + eval(b.right());
                case '-' -> eval(b.left()) - eval(b.right());
                case '*' -> eval(b.left()) * eval(b.right());
                case '/' -> eval(b.left()) / eval(b.right());
                default  -> throw new IllegalStateException("op " + b.op());
            };
        };
    }

    public static void main(String[] args) {
        List<String> cases = new ArrayList<>(List.of(
            "1 + 2",
            "2 * 3 + 4",
            "(1 + 2) * 3",
            "-5 + 3"
        ));
        long[] expected = { 3L, 10L, 9L, -2L };
        for (int i = 0; i < cases.size(); i++) {
            long got = eval(new Parser(cases.get(i)).parseExpr(0));
            System.out.printf("%-15s => %d (expected %d)%n",
                cases.get(i), got, expected[i]);
            assert got == expected[i] : "mismatch at " + cases.get(i);
        }
    }
}

// 2.3.2 error-hole coverage: a `\u` escape inside a char
// literal. ADR 0010's char scanner models 'C', '\C' and
// '\xHH', so the four-hex form needs the operator fallback.
class UnicodeEscapes {
    static final char A = '\u0041';
    static final char NL = '\n';
}

File: tests/corpus/concept.php
<?php
// vyakarana stand-in -- not from vidya. See docs/adr/0006-standin-corpus-policy.md.
// Re-sync when vidya adds a PHP reference sample.
//
// Lexing + parsing of arithmetic expressions, modern PHP.

declare(strict_types=1);

namespace Vyakarana\Concept;

/*
 * Multi-line block comment exercising the slash-star pair rule
 * across multiple lines. Same shape as C / Java / C++.
 */

// ── Tokens ───────────────────────────────────────────────────────────

enum TokenKind: string {
    case Number = 'number';
    case Plus   = 'plus';
    case Minus  = 'minus';
    case Star   = 'star';
    case Slash  = 'slash';
    case LParen = 'lparen';
    case RParen = 'rparen';
    case Eof    = 'eof';
}

final readonly class Token {
    public function __construct(
        public TokenKind $kind,
        public string $text,
        public int $pos,
    ) {}

    public function __toString(): string {
        return "Token({$this->kind->value}, {$this->text}, {$this->pos})";
    }
}

// ── Lexer ────────────────────────────────────────────────────────────

final class Lexer {
    private int $pos = 0;

    public function __construct(private readonly string $source) {}

    private function skipWhitespace(): void {
        while ($this->pos < strlen($this->source) && ctype_space($this->source[$this->pos])) {
            $this->pos++;
        }
    }

    public function next(): Token {
        $this->skipWhitespace();
        $start = $this->pos;

        if ($this->pos >= strlen($this->source)) {
            return new Token(TokenKind::Eof, '', $this->pos);
        }

        $c = $this->source[$this->pos];
        switch ($c) {
            case '+': $this->pos++; return new Token(TokenKind::Plus,   '+', $start);
            case '-': $this->pos++; return new Token(TokenKind::Minus,  '-', $start);
            case '*': $this->pos++; return new Token(TokenKind::Star,   '*', $start);
            case '/': $this->pos++; return new Token(TokenKind::Slash,  '/', $start);
            case '(': $this->pos++; return new Token(TokenKind::LParen, '(', $start);
            case ')': $this->pos++; return new Token(TokenKind::RParen, ')', $start);
        }

        if (ctype_digit($c)) {
            while ($this->pos < strlen($this->source) && ctype_digit($this->source[$this->pos])) {
                $this->pos++;
            }
            return new Token(TokenKind::Number, substr($this->source, $start, $this->pos - $start), $start);
        }

        throw new \RuntimeException("unexpected character at {$this->pos}");
    }
}

// ── AST ──────────────────────────────────────────────────────────────

abstract class Expr {
    abstract public function eval(): int;
}

final class NumberExpr extends Expr {
    public function __construct(public readonly int $value) {}
    public function eval(): int { return $this->value; }
}

final class UnaryExpr extends Expr {
    public function __construct(
        public readonly string $op,
        public readonly Expr $operand,
    ) {}
    public function eval(): int { return -$this->operand->eval(); }
}

final class BinOpExpr extends Expr {
    public function __construct(
        public readonly string $op,
        public readonly Expr $left,
        public readonly Expr $right,
    ) {}

    public function eval(): int {
        $a = $this->left->eval();
        $b = $this->right->eval();
        return match ($this->op) {
            '+' => $a + $b,
            '-' => $a - $b,
            '*' => $a * $b,
            '/' => intdiv($a, $b),
            default => throw new \RuntimeException("bad op {$this->op}"),
        };
    }
}

// ── Pratt parser ─────────────────────────────────────────────────────

final class Parser {
    private Lexer $lexer;
    private Token $current;

    public function __construct(string $source) {
        $this->lexer = new Lexer($source);
        $this->current = $this->lexer->next();
    }

    private function advance(): void {
        $this->current = $this->lexer->next();
    }

    private static function infixBp(TokenKind $k): int {
        return match ($k) {
            TokenKind::Plus, TokenKind::Minus => 1,
            TokenKind::Star, TokenKind::Slash => 3,
            default => 0,
        };
    }

    public function parseExpr(int $minBp): Expr {
        $left = $this->parsePrimary();
        while (true) {
            $bp = self::infixBp($this->current->kind);
            if ($bp === 0 || $bp < $minBp) break;
            $op = $this->current->text;
            $this->advance();
            $right = $this->parseExpr($bp + 1);
            $left = new BinOpExpr($op, $left, $right);
        }
        return $left;
    }

    private function parsePrimary(): Expr {
        if ($this->current->kind === TokenKind::Minus) {
            $this->advance();
            return new UnaryExpr('-', $this->parsePrimary());
        }
        if ($this->current->kind === TokenKind::LParen) {
            $this->advance();
            $inner = $this->parseExpr(0);
            if ($this->current->kind !== TokenKind::RParen) {
                throw new \RuntimeException('expected )');
            }
            $this->advance();
            return $inner;
        }
        if ($this->current->kind === TokenKind::Number) {
            $v = (int) $this->current->text;
            $this->advance();
            return new NumberExpr($v);
        }
        throw new \RuntimeException("unexpected token {$this->current}");
    }
}

// ── Demo ─────────────────────────────────────────────────────────────

$cases = [
    ['1 + 2',       3],
    ['2 * 3 + 4',   10],
    ['(1 + 2) * 3', 9],
    ['-5 + 3',      -2],
];

foreach ($cases as [$input, $expected]) {
    $got = (new Parser($input))->parseExpr(0)->eval();
    printf("%-15s => %d (expected %d)\n", $input, $got, $expected);
    if ($got !== $expected) {
        throw new \RuntimeException("mismatch at \"{$input}\"");
    }
}

<?php
// 2.3.2 error-hole coverage: PHP's backtick operator is
// shell_exec() by another spelling.
$listing = `ls -l`;
echo $listing;
